en/lessons/basics/containers.md

```diff
@@ -0,0 +1,799 @@
+
+```haskell
+ghci> :{
+| safeHead :: [a] -> Maybe a
+| safeHead [] = Nothing
+| safeHead (x:xs) = Just x
+|
+| safeTail :: [a] -> Maybe [a]
+| safeTail [] = Nothing
+| safeTail (x:xs) = Just xs
+| :}
+ghci> safeHead ["Orange", "Banana", "Apple"]
+Just "Orange"
+ghci> safeHead []
+Nothing
+ghci> safeTail ["Orange", "Banana", "Apple"]
+Just ["Banana","Apple"]
+ghci> safeTail []
+Nothing
+```
+__Note__: `:{` and `:}` allow you to write multiline definitions in ghci.
+
+Hooray! No more exceptions.
+
+Another container that is the cousin of the list, and ensures that `head` and
+`tail` are safe is the nonempty list.
+
+```haskell
+ghci> :i NE.NonEmpty
+type NE.NonEmpty :: * -> *
+data NE.NonEmpty a = a NE.:| [a]
+  	-- Defined in ‘GHC.Base’
+```
+
+From its definition we can see that `NonEmpty` requires the first element to be
+present. This symbol `:|` is a constructor just like `:`, in fact the definition of
+`NonEmpty` is identical to that of lists, except it omits the `[]` (Nil) case.
+
+This handles the partiality problem the exact opposite way as the
+`Maybe` solution. Instead of forcing the caller of the function to guard against
+the ill-defined case when handling the result of the function, it forces them to
+construct a valid input up front (when calling the function).
+
+```haskell
+ghci> :t NE.head
+NE.head :: NE.NonEmpty a -> a
+ghci> NE.head (1 NE.:| [])
+1
+ghci> NE.head []
+
+<interactive>:11:9: error:
+    • Couldn't match expected type: NE.NonEmpty a
+                  with actual type: [a0]
+    ...
+```
+
+Notice that this time the error is not a runtime exception but a type error, the
+compiler is telling us that we tried to use a (potentially empty) list rather
+than the required `NonEmpty` type.
+
+### List Performance
+
+Haskell implements lists as linked lists. The cons cells (the operator `:` is
+called cons, short for constructor) act as the links. This dictates which
+operations can be done quickly and which can be slow:
+
+Prepending a value to a list is easy and fast - all we have to do is create a
+new cons cell with the element we want to prepend and point it to the existing
+list.
+
+```haskell
+prepend value list = value : list
+```
+
+On the other hand, since the list data type (as shown above) can be either
+empty (`[]` or `Nil`) or a cons cell that will point to the rest of the list,
+it does not contain information about the length of the list, or a reference to
+the end of the list.
+
+Because of that, in order to retrieve the length of a list we must walk
+each cons cell and count until we reach the end of the list. To find the
+value at a specific index we need to traverse the list until we reach it.
+
+Similarly, in order to append a list to an existing list, we need to go to the
+end of the existing list, and add a cons cell that points to the new list:
+
+```haskell
+append originalList newList =
+    case originalList of
+        [] -> newList
+        x : xs -> x : append xs newList
+```
+
```

Comment:
There should be a mention of `++` in particular, I think: its usage really is a code smell for anything more complex than hello ++ world. In particular, I’d mention as an example that having ++ in a loop brings you quadratic performance.